### ISSUE TITLE
[Question][HW2] Stack overflow when irgen fuzzing.

### ISSUE THREAD

**Issue by leejy12** [kaist-cp/cs420#483]
I am grading my solution for homework 2.

After running the fuzzer, it found a test case that failed. 
However, the generated IR from my solution and the model solution at gg are **identical**.

How can I resolve this error?

For context, here's the `test_reduced.c`:
```c
int a;
int main() {
  int b;
  for (; a < 6; a++)
    if (b)
      b = 0;
}
```

Here's the generated IR
```
var i32 @a = default

fun i32 @main () {
init:
  bid: b0
  allocations: 
    %l0:i32:b

block b0:
  j b1()

block b1:
  j b2()

block b2:
  %b2:i0:i32 = load @a:i32*
  %b2:i1:u1 = cmp lt %b2:i0:i32 6:i32
  br %b2:i1:u1, b3(), b5()

block b3:
  %b3:i0:i32 = load %l0:i32*
  %b3:i1:u1 = cmp ne %b3:i0:i32 0:i32
  br %b3:i1:u1, b6(), b7()

block b4:
  %b4:i0:i32 = load @a:i32*
  %b4:i1:i32 = add %b4:i0:i32 1:i32
  %b4:i2:unit = store %b4:i1:i32 @a:i32*
  j b2()

block b5:
  ret 0:i32

block b6:
  %b6:i0:unit = store 0:i32 %l0:i32*
  j b8()

block b7:
  j b8()

block b8:
  j b4()
}
```

**Comment by leejy12** [comment]
I think the problem is that my solution produces a stack overflow. I'll noodle more...

**Comment by AnHaechan** [comment]
The grade is determined by running [grade-irgen.sh](https://github.com/kaist-cp/kecc-public/blob/main/scripts/grade-irgen.sh).
That is, getting stack overflow from running `RUST_MIN_STACK=33554432 python3 tests/fuzz.py --irgen -n80 --seed 22` ([Line 12 from grade-irgen.sh](https://github.com/kaist-cp/kecc-public/blob/main/scripts/grade-irgen.sh#L12)) means that your solution needs to be modified.
I suspect `rust-lldb` may use larger stack size on its own, but that is orthogonal to our grading...

**Comment by leejy12** [comment]
Here's what I found so far.

Running `rust-lldb target/release/kecc` works just fine in the debugger.
However, if I run `rust-lldb target/debug/kecc` and run again, I get a segmentation fault!

```
$ rust-lldb target/debug/kecc
(lldb) run -i tests/test_polished.c
Process 411 launched: '/home/leejy/kecc/target/debug/kecc' (x86_64)
Process 411 stopped and restarted: thread 1 received signal: SIGCHLD
Process 411 stopped
* thread #1, name = 'kecc', stop reason = signal SIGSEGV: invalid address (fault address: 0x7fffff7fee10)
    frame #0: 0x00005555559f045e kecc`__rust_probestack + 23
kecc`__rust_probestack:
->  0x5555559f045e <+23>: testq  %rsp, 0x8(%rsp)
    0x5555559f0463 <+28>: subq   $0x1000, %r11             ; imm = 0x1000 
    0x5555559f046a <+35>: cmpq   $0x1000, %r11             ; imm = 0x1000 
    0x5555559f0471 <+42>: ja     0x5555559f0457            ; <+16>
(lldb) bt
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
* thread #1, name = 'kecc', stop reason = signal SIGSEGV: invalid address (fault address: 0x7fffff7fee10)
  * frame #0: 0x00005555559f045e kecc`__rust_probestack + 23
    frame #1: 0x00005555556e829a kecc`kecc::irgen::IrgenFunc::translate_binary_op::h1f93570bc71a4d16(self=0x0000000000000000, op=Index, lhs=0x0000000000000000, rhs=0x0000000000000000, context=0x0000000000000000) at mod.rs:1989
```

At the last call site, it seems that `translate_binary_op` is called with null arguments. I have no idea how this happened. Also, given how my program behaves differently between debug and release, it might have undefined behavior.

I'll keep investigating...

**Comment by leejy12** [comment]
I found the culprit.

Let's see this C code:
```c
int main()
{
    unsigned int u = 420;
    if (u)
    {
        return 1;
    }
}
```

When evaluating the condition `if (u)`, I was not respecting the signedness of `u`. So I was comparing it with `0i32` not `0u32`. This causes the IR interpreter to trigger an assert. I still have no idea why the stack was blowing up...

After solving the above issue, I made some more progress in the fuzzer, so I'll close this issue.
